OTP-25.3 segmentation fault running Cowboy tests on Linux/FreeBSD

**Describe the bug**
I am getting a segfault running the Cowboy test suite with OTP-25.3 on Linux. I am seeing the error with the Arch Linux package as well as with builds using kerl. At the time of writing I have seen it on an Arch Linux laptop, Arch Linux LXC container running on an Arch Linux server (both a bit old) and Debian and Ubuntu LXC containers on another Arch Linux server.

The problem does not seem to occur on macOS.

**To Reproduce**
```
git clone https://github.com/ninenines/cowboy
cd cowboy
make ct
```

**Expected behavior**
No segfault!

**Affected versions**
OTP-25.3 only (no problem with master or 25.2).

**Additional context**
Example session:

```
essen@zen 0 % git clone https://github.com/ninenines/cowboy
Cloning into 'cowboy'...
remote: Enumerating objects: 13367, done.
remote: Counting objects: 100% (165/165), done.
remote: Compressing objects: 100% (99/99), done.
remote: Total 13367 (delta 72), reused 136 (delta 59), pack-reused 13202
Receiving objects: 100% (13367/13367), 8.21 MiB | 2.92 MiB/s, done.
Resolving deltas: 100% (8790/8790), done.
essen@zen 0 % cd cowboy 
essen@zen 0 % make ct
 DEP    ci.erlang.mk (master)
 GEN    clean-app
 GEN    coverdata-clean
 DEP    cowlib (2.12.1)
 DEP    ranch (1.8.0)
make[1]: Entering directory '/home/essen/tmp/cowboy/deps/cowlib'
 DEPEND cowlib.d
 ERLC   cow_base64url.erl cow_cookie.erl cow_date.erl cow_hpack.erl cow_http.erl cow_http2.erl cow_http2_machine.erl cow_http_hd.erl cow_http_struct_hd.erl cow_http_te.erl cow_iolists.erl cow_link.erl cow_mimetypes.erl cow_multipart.erl cow_qs.erl cow_spdy.erl cow_sse.erl cow_uri.erl cow_uri_template.erl cow_ws.erl
 APP    cowlib
make[1]: Leaving directory '/home/essen/tmp/cowboy/deps/cowlib'
make[1]: Entering directory '/home/essen/tmp/cowboy/deps/ranch'
 DEPEND ranch.d
 ERLC   ranch.erl ranch_acceptor.erl ranch_acceptors_sup.erl ranch_app.erl ranch_conns_sup.erl ranch_crc32c.erl ranch_listener_sup.erl ranch_protocol.erl ranch_proxy_header.erl ranch_server.erl ranch_ssl.erl ranch_sup.erl ranch_tcp.erl ranch_transport.erl
 APP    ranch
make[1]: Leaving directory '/home/essen/tmp/cowboy/deps/ranch'
 DEP    ct_helper (master)
 DEP    gun (master)
make[1]: Entering directory '/home/essen/tmp/cowboy/deps/ct_helper'
 DEPEND ct_helper.d
 ERLC   ct_helper.erl ct_helper_error_h.erl
 APP    ct_helper.app.src
make[1]: Leaving directory '/home/essen/tmp/cowboy/deps/ct_helper'
make[1]: Entering directory '/home/essen/tmp/cowboy/deps/gun'
 DEPEND gun.d
 ERLC   gun.erl gun_app.erl gun_conns_sup.erl gun_content_handler.erl gun_cookies.erl gun_cookies_list.erl gun_data_h.erl gun_default_event_h.erl gun_event.erl gun_http.erl gun_http2.erl gun_pool.erl gun_pool_events_h.erl gun_pools_sup.erl gun_protocols.erl gun_public_suffix.erl gun_raw.erl gun_socks.erl gun_sse_h.erl gun_sup.erl gun_tcp.erl gun_tcp_proxy.erl gun_tls.erl gun_tls_proxy.erl gun_tls_proxy_cb.erl gun_tls_proxy_http2_connect.erl gun_tunnel.erl gun_ws.erl gun_ws_h.erl gun_ws_protocol.erl
 APP    gun
 GEN    rebar.config
make[1]: Leaving directory '/home/essen/tmp/cowboy/deps/gun'
 DEPEND cowboy.d
 ERLC   cowboy.erl cowboy_app.erl cowboy_bstr.erl cowboy_children.erl cowboy_clear.erl cowboy_clock.erl cowboy_compress_h.erl cowboy_constraints.erl cowboy_handler.erl cowboy_http.erl cowboy_http2.erl cowboy_loop.erl cowboy_metrics_h.erl cowboy_middleware.erl cowboy_req.erl cowboy_rest.erl cowboy_router.erl cowboy_static.erl cowboy_stream.erl cowboy_stream_h.erl cowboy_sub_protocol.erl cowboy_sup.erl cowboy_tls.erl cowboy_tracer_h.erl cowboy_websocket.erl
src/cowboy_http.erl:695:1: Warning: function horse_clean_value_ws_end/0 is unused
%  695| horse_clean_value_ws_end() ->
%     | ^

 APP    cowboy
 GEN    test-build
 DEPEND cowboy.d
 ERLC   compress_SUITE.erl cowboy_ct_hook.erl cowboy_test.erl examples_SUITE.erl h2spec_SUITE.erl accept_callback_h.erl accept_callback_missing_h.erl asterisk_h.erl charset_in_content_types_provided_h.erl charset_in_content_types_provided_implicit_h.erl charset_in_content_types_provided_implicit_no_callback_h.erl charsets_provided_empty_h.erl charsets_provided_h.erl compress_h.erl content_types_accepted_h.erl content_types_provided_h.erl crash_h.erl create_resource_h.erl custom_req_fields_h.erl default_h.erl delay_hello_h.erl delete_resource_h.erl echo_h.erl expires_h.erl generate_etag_h.erl hello_h.erl if_range_h.erl last_modified_h.erl long_polling_h.erl long_polling_sys_h.erl loop_handler_abort_h.erl loop_handler_body_h.erl loop_handler_timeout_h.erl multipart_h.erl provide_callback_missing_h.erl provide_range_callback_h.erl range_satisfiable_h.erl ranges_provided_auto_h.erl ranges_provided_h.erl rate_limited_h.erl resp_h.erl resp_iolist_body_h.erl rest_hello_h.erl send_message_h.erl set_options_h.erl stop_handler_h.erl stream_handler_h.erl switch_handler_h.erl switch_protocol_flush_h.erl ws_active_commands_h.erl ws_deflate_commands_h.erl ws_deflate_opts_h.erl ws_dont_validate_utf8_h.erl ws_handle_commands_h.erl ws_info_commands_h.erl ws_init_commands_h.erl ws_init_h.erl ws_set_options_commands_h.erl ws_shutdown_reason_commands_h.erl ws_terminate_h.erl http2_SUITE.erl http_SUITE.erl loop_handler_SUITE.erl metrics_SUITE.erl misc_SUITE.erl plain_handler_SUITE.erl proxy_header_SUITE.erl req_SUITE.erl rest_handler_SUITE.erl rfc6585_SUITE.erl rfc7230_SUITE.erl rfc7231_SUITE.erl rfc7538_SUITE.erl rfc7540_SUITE.erl rfc8297_SUITE.erl rfc8441_SUITE.erl security_SUITE.erl static_handler_SUITE.erl stream_handler_SUITE.erl sys_SUITE.erl tracer_SUITE.erl ws_SUITE.erl ws_echo.erl ws_echo_timer.erl ws_init_shutdown.erl ws_max_frame_size.erl ws_send_many.erl ws_subprotocol.erl ws_timeout_cancel.erl ws_timeout_hibernate.erl ws_autobahn_SUITE.erl ws_handler_SUITE.erl
 GEN    /home/essen/tmp/cowboy/.erlang.mk/gopath/src/github.com/summerwind/h2spec/h2spec
Cloning into '/home/essen/tmp/cowboy/.erlang.mk/gopath/src/github.com/summerwind/h2spec'...
remote: Enumerating objects: 129, done.
remote: Counting objects: 100% (129/129), done.
remote: Compressing objects: 100% (118/118), done.
remote: Total 129 (delta 50), reused 44 (delta 7), pack-reused 0
Receiving objects: 100% (129/129), 70.20 KiB | 1.95 MiB/s, done.
Resolving deltas: 100% (50/50), done.
make[1]: Entering directory '/home/essen/tmp/cowboy/.erlang.mk/gopath/src/github.com/summerwind/h2spec'
go build -ldflags "-X main.VERSION=2.6.0 -X main.COMMIT=af83a65f0b6273ef38bf778d400d98892e7653d8" cmd/h2spec/h2spec.go
go: downloading github.com/spf13/cobra v0.0.0-20170118185516-dc208f4211e7
go: downloading golang.org/x/net v0.0.0-20161104230106-55a3084c9119
go: downloading github.com/fatih/color v0.0.0-20161025120501-bf82308e8c85
go: downloading github.com/mattn/go-isatty v0.0.4
go: downloading github.com/mattn/go-colorable v0.1.0
go: downloading github.com/spf13/pflag v1.0.3
make[1]: Leaving directory '/home/essen/tmp/cowboy/.erlang.mk/gopath/src/github.com/summerwind/h2spec'
 GEN    ct


Common Test v1.24 starting (cwd is /home/essen/tmp/cowboy)



CWD set to: "/home/essen/tmp/cowboy/logs/ct_run.ct_cowboy@zen.2023-04-03_11.52.38"

TEST INFO: 25 test(s), 4807 case(s) in 25 suite(s)

Testing tmp.cowboy.compress_SUITE: Starting test, 84 test cases
make: *** [erlang.mk:6679: ct] Segmentation fault (core dumped)
essen@zen 2 % coredumpctl debug
           PID: 706462 (beam.smp)
           UID: 1000 (essen)
           GID: 1000 (essen)
        Signal: 11 (SEGV)
     Timestamp: Mon 2023-04-03 11:52:40 CEST (7s ago)
  Command Line: ct_run -- -root /usr/lib/erlang -bindir /usr/lib/erlang/erts-13.2/bin -progname erl -- -home /home/essen -- -sname ct_cowboy -s ct_run script_start -s erlang halt -no_auto_compile -noshell -noinput -pa /home/essen/tmp/cowboy/ebin /home/essen/tmp/cowboy/test -dir /home/essen/tmp/cowboy/test -logdir /home/essen/tmp/cowboy/logs -suite compress_SUITE h2spec_SUITE http_SUITE http2_SUITE loop_handler_SUITE metrics_SUITE misc_SUITE plain_handler_SUITE proxy_header_SUITE req_SUITE rest_handler_SUITE rfc6585_SUITE rfc7230_SUITE rfc7231_SUITE rfc7538_SUITE rfc7540_SUITE rfc8297_SUITE rfc8441_SUITE security_SUITE static_handler_SUITE stream_handler_SUITE sys_SUITE tracer_SUITE ws_SUITE ws_handler_SUITE -ct_hooks cowboy_ct_hook $'[]'
    Executable: /usr/lib/erlang/erts-13.2/bin/beam.smp
 Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/vte-spawn-dd4f91cc-fb32-48f5-b894-27d42b3e871d.scope
          Unit: user@1000.service
     User Unit: vte-spawn-dd4f91cc-fb32-48f5-b894-27d42b3e871d.scope
         Slice: user-1000.slice
     Owner UID: 1000 (essen)
       Boot ID: 96a8acde07194f84b68c5e65a31c7f3d
    Machine ID: 811435a01a6e4a3286918c9c9c5daa72
      Hostname: zen
       Storage: /var/lib/systemd/coredump/core.beam\x2esmp.1000.96a8acde07194f84b68c5e65a31c7f3d.706462.1680515560000000.zst (present)
  Size on Disk: 30.4M
       Message: Process 706462 (beam.smp) of user 1000 dumped core.
                
                Stack trace of thread 706477:
                #0  0x000055d6f863d4a3 n/a (beam.smp + 0x2934a3)
                #1  0x000055d6f864207f erts_handle_canceled_timers (beam.smp + 0x29807f)
                #2  0x000055d6f841a591 n/a (beam.smp + 0x70591)
                #3  0x000055d6f8430011 erts_schedule (beam.smp + 0x86011)
                #4  0x00007fa611500625 n/a (n/a + 0x0)
                ELF object binary architecture: AMD x86-64

GNU gdb (GDB) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lib/erlang/erts-13.2/bin/beam.smp...

This GDB supports auto-downloading debuginfo from the following URLs:
https://debuginfod.archlinux.org 
Enable debuginfod for this session? (y or [n]) y
Debuginfod has been enabled.
To make this setting permanent, add 'set debuginfod enabled on' to .gdbinit.
Reading symbols from /home/essen/.cache/debuginfod_client/fadb3b2cbfdac3ebcab31101e921630ee005fc2d/debuginfo...

warning: Can't open file /memfd:vmem (deleted) during file-backed mapping note processing
[New LWP 706477]
[New LWP 706502]
[New LWP 706484]
[New LWP 706462]
[New LWP 706470]
[New LWP 706501]
[New LWP 706485]
[New LWP 706488]
[New LWP 706493]
[New LWP 706506]
[New LWP 706508]
[New LWP 706512]
[New LWP 706494]
[New LWP 706490]
[New LWP 706511]
[New LWP 706476]
[New LWP 706491]
[New LWP 706468]
[New LWP 706492]
[New LWP 706480]
[New LWP 706500]
[New LWP 706475]
[New LWP 706481]
[New LWP 706478]
[New LWP 706495]
[New LWP 706515]
[New LWP 706486]
[New LWP 706469]
[New LWP 706487]
[New LWP 706504]
[New LWP 706498]
[New LWP 706472]
[New LWP 706505]
[New LWP 706474]
[New LWP 706509]
[New LWP 706482]
[New LWP 706510]
[New LWP 706503]
[New LWP 706479]
[New LWP 706483]
[New LWP 706497]
[New LWP 706513]
[New LWP 706473]
[New LWP 706499]
[New LWP 706489]
[New LWP 706496]
[New LWP 706507]
[New LWP 706514]

warning: Section `.reg-xstate/706477' in core file too small.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
JITed symbol file is not an object file, ignoring it.
JITed symbol file is not an object file, ignoring it.
JITed symbol file is not an object file, ignoring it.
(...)
JITed symbol file is not an object file, ignoring it.
--Type <RET> for more, q to quit, c to continue without paging--
Core was generated by `ct_run -- -root /usr/lib/erlang -bindir /usr/lib/erlang/erts-13.2/bin -progname'.
Program terminated with signal SIGSEGV, Segmentation fault.

warning: Section `.reg-xstate/706477' in core file too small.
#0  0x000055d6f863d4a3 in btm_rbt_delete (root=0x7fa6104eda38, n=n@entry=0x7fa614347be0)
    at beam/erl_rbtree.h:723
Downloading 0.05 MB source file /usr/src/debug/erlang/otp/erts/emulator/beam/erl_rbtree.h
723                 ERTS_RBT_T *t = ERTS_RBT_GET_LEFT(y);                                            
[Current thread is 1 (Thread 0x7fa6103eb6c0 (LWP 706477))]
(gdb) bt
#0  0x000055d6f863d4a3 in btm_rbt_delete (root=0x7fa6104eda38, n=n@entry=0x7fa614347be0)
    at beam/erl_rbtree.h:723
#1  0x000055d6f864207f in cleanup_sched_local_canceled_timer (tmr=0x7fa614347be0, 
    esdp=0x7fa612b3bfc0) at beam/erl_hl_timer.c:1416
#2  handle_canceled_queue (use_limit=1, ops_limit=100, need_more_work=<optimized out>, 
    thr_prgr_p=<optimized out>, need_thr_progress=0x7fa6103e6b78, cq=0x7fa6104ed9c0, 
    esdp=0x7fa612b3bfc0) at beam/erl_hl_timer.c:1606
#3  erts_handle_canceled_timers (vesdp=0x7fa612b3bfc0, need_thr_progress=0x7fa6103e6b78, 
    thr_prgr_p=0x7fa6103e6b80, need_more_work=0x7fa6103e6b7c) at beam/erl_hl_timer.c:1647
#4  0x000055d6f841a591 in handle_canceled_timers_thr_prgr (waiting=0, aux_work=131, 
    awdp=0x7fa612b3c0b8) at beam/erl_process.c:2234
#5  handle_aux_work (awdp=0x7fa612b3c0b8, orig_aux_work=<optimized out>, waiting=0)
    at beam/erl_process.c:2699
#6  0x000055d6f8430011 in erts_schedule (esdp=<optimized out>, p=<optimized out>, 
    calls=<optimized out>) at beam/erl_process.c:9693
#7  0x00007fa611500625 in ?? ()
#8  0x0000000000000000 in ?? ()
(gdb) q
essen@zen 0 % pacman -Qi erlang
Name            : erlang
Version         : 25.3-1
```

--- a/erts/emulator/beam/code_ix.c
+++ b/erts/emulator/beam/code_ix.c
@@ -347,7 +347,17 @@ static ErtsThrPrgrLaterOp global_code_barrier_lop;
 
 static void decrement_blocking_code_barriers(void *ignored) {
     (void)ignored;
-    erts_atomic32_dec_nob(&outstanding_blocking_code_barriers);
+
+    if (erts_atomic32_dec_read_nob(&outstanding_blocking_code_barriers) > 0) {
+        /* We had more than one barrier in the same tick, and can't tell
+         * whether the later ones were issued before any of the managed threads
+         * were woken. Keep telling all managed threads to execute an
+         * instruction barrier on wake-up for one more tick. */
+        erts_atomic32_set_nob(&outstanding_blocking_code_barriers, 1);
+        erts_schedule_thr_prgr_later_op(decrement_blocking_code_barriers,
+                                        NULL,
+                                        &global_code_barrier_lop);
+    }
 }
 
 static void schedule_blocking_code_barriers(void *ignored) {
@@ -358,11 +368,12 @@ static void schedule_blocking_code_barriers(void *ignored) {
      * counter.
      *
      * Note that we increment and decrement instead of setting and clearing
-     * since we might execute several blocking barriers in the same tick. */
-    erts_atomic32_inc_nob(&outstanding_blocking_code_barriers);
-    erts_schedule_thr_prgr_later_op(decrement_blocking_code_barriers,
-                                    NULL,
-                                    &global_code_barrier_lop);
+     * since we might schedule several blocking barriers in the same tick. */
+    if (erts_atomic32_inc_read_nob(&outstanding_blocking_code_barriers) == 1) {
+        erts_schedule_thr_prgr_later_op(decrement_blocking_code_barriers,
+                                        NULL,
+                                        &global_code_barrier_lop);
+    }
 }
 #endif
 

